core/src/apps/solana/transaction/instruction.py

```diff
@@ -0,0 +1,141 @@
+    instruction_id: int
+
+    property_templates: list[PropertyTemplate]
+    accounts_template: list[AccountTemplate]
+
+    ui_name: str
+
+    ui_properties: list[UIProperty]
+
+    parsed_data: dict[str, Any]
+    parsed_accounts: dict[str, Account]
+
+    is_program_supported: bool
+    is_instruction_supported: bool
+    instruction_data: bytes
+    accounts: list[Account]
+
+    multisig_signers: list[Account]
+
+    is_deprecated_warning: str | None = None
+
+    @staticmethod
+    def parse_instruction_data(
+        instruction_data: bytes, property_templates: list[PropertyTemplate]
+    ):
+        from trezor.utils import BufferReader
+        from trezor.wire import DataError
+
+        reader = BufferReader(instruction_data)
+
+        parsed_data = {}
+        for property_template in property_templates:
+            property = property_template.parse(reader)
+
+            parsed_data[property_template.name] = property
+
+        if reader.remaining_count() != 0:
+            raise DataError("Invalid transaction data")
+
+        return parsed_data
+
+    @staticmethod
+    def parse_instruction_accounts(
+        accounts: list[Account], accounts_template: list[AccountTemplate]
+    ):
+        parsed_account = {}
+        for i, account_template in enumerate(accounts_template):
+            if i >= len(accounts):
+                if account_template.optional:
+                    continue
+                else:
+                    raise ValueError  # "Account is missing
+
+            parsed_account[account_template.name] = accounts[i]
+        return parsed_account
+
+    def __init__(
+        self,
+        instruction_data: bytes,
+        program_id: str,
+        accounts: list[Account],
+        instruction_id: int,
+        property_templates: list[PropertyTemplate],
+        accounts_template: list[AccountTemplate],
+        ui_properties: list[UIProperty],
+        ui_name: str,
+        is_program_supported: bool = True,
+        is_instruction_supported: bool = True,
+        supports_multisig: bool = False,
+        is_deprecated_warning: str | None = None,
+    ) -> None:
+        self.program_id = program_id
+        self.instruction_id = instruction_id
+
+        self.property_templates = property_templates
+        self.accounts_template = accounts_template
+
+        self.ui_name = ui_name
+
+        self.ui_properties = ui_properties
+
+        self.is_program_supported = is_program_supported
+        self.is_instruction_supported = is_instruction_supported
+
+        self.is_deprecated_warning = is_deprecated_warning
+
+        self.instruction_data = instruction_data
+        self.accounts = accounts
+
+        if self.is_instruction_supported:
+            self.parsed_data = self.parse_instruction_data(
+                instruction_data, property_templates
+            )
+
+            self.parsed_accounts = self.parse_instruction_accounts(
+                accounts, accounts_template
+            )
```

Comment:
these must not be conditional, otherwise you're not initializing `parsed_data` and `parsed_accounts`. the functions parse_instructions_x will cleanly handle the empty list of property/account templates, no?

Reply:
In `parse_instruction_data` we verify that the whole buffer was used. This can't be done with unknown instructions so no, the functions don't handle empty templates cleanly. I think the best way to approach is simply add an `else` to the whole block which would initialize all three attributes - https://github.com/trezor/trezor-firmware/pull/3359/commits/853af2f57fa017c5ae628344d4e1363c68936664#diff-8494caa0738cc8ad43c528cd82094529785133f6a2d7c4747c6ab7a91b6a034eR114.